proposals/interface-types/working-notes/Types.md

```diff
@@ -37,36 +62,158 @@ Integer values.
 | u32  | [0, 2^32-1]       |
 | u64  | [0, 2^64-1]       |
 
-These types differ from the core value types in that they represent a range of integers instead of a set of bits.
 
-## `string`
+```
+integraltype ::= s8 | u8 | s16 | u16 | s32 | u32 | s64 | u64
+```
+
+Integral types denote an integer value in a given range. There is no implied
+storage layout.
 
-A string value is a sequence of [unicode code points](https://www.unicode.org/glossary/#code_point).
+#### String Type
 
-## `record`
+The `string` type denotes a sequence of
+[unicode code points](https://www.unicode.org/glossary/#code_point).
 
-The `record` interface value type is a compound type defined by the following abstract syntax:
 
 ```
-#record ::= record $fields: (field #interface-valtype)*
+stringtype ::= string
 ```
 
-A `record` value is an ordered set of `|$fields|` interface values, where each interface value is of the corresponding type in `$fields`. This is sometimes known as a struct or product type.
+>Note: In a future revision, this may be reinterpreted as a macro of a more
+>explicit designation of the `sequence` type.
+
+Note that there is no implied representation or encoding implied in this.
+
 
-A `record` type is valid iff:
- * `|$fields| > 0`
- * For every `field` in `$fields`
-  * `#interface-valtype` is valid
+### Structured Value Types
 
-## `array`
+Structured types allow the representation of more complex combinations of
+values. There are five forms of structured type: arrays, sequences, records,
+protocols and algebraic variants.
 
-The `array` interface value type is a compound type defined by the following abstract syntax:
+```
+structuretype ::= optiontype | eithertype |
+   arraytype | sequencetype | recordtype | varianttype
+```
+
+#### Option Type
+
+Option types are used to model nullability in types.
 
 ```
-#array ::= array #interface-valtype
+optiontype ::= option type
 ```
 
-An `array` value is a homogenous sequence of interface values with type `$interface-value-type`.
+The `option` type can be viewed as a use of type variants -- with one variant
+being the enumerated symbol `$none` and the other being the single value wrapped
+in a `$some` variant.
+
+#### Either Type
+
+The `either` type is used to model situations where one of two values may be
+returned by a function. A classic case for this is in modeling error returns in
+APIs.
+
+```
+eithertype ::= either type
```

Comment:
I think there should be two `type`'s as part of this production for the two possible values an either may have.